src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/DecryptorPalWindows.Decrypt.cs

```diff
@@ -0,0 +1,223 @@
+
+            return null;
+        }
+
+        private Exception TryDecryptAgree(KeyAgreeRecipientInfo keyAgreeRecipientInfo, SafeProvOrNCryptKeyHandle hKey, CryptKeySpec keySpec, X509Certificate2Collection originatorCerts, X509Certificate2Collection extraStore)
+        {
+            unsafe
+            {
+                KeyAgreeRecipientInfoPalWindows pal = (KeyAgreeRecipientInfoPalWindows)(keyAgreeRecipientInfo.Pal);
+                return pal.WithCmsgCmsRecipientInfo<Exception>(
+                    delegate (CMSG_KEY_AGREE_RECIPIENT_INFO* pKeyAgreeRecipientInfo)
+                    {
+                        CMSG_CTRL_KEY_AGREE_DECRYPT_PARA decryptPara = default(CMSG_CTRL_KEY_AGREE_DECRYPT_PARA);
+                        decryptPara.cbSize = Marshal.SizeOf<CMSG_CTRL_KEY_AGREE_DECRYPT_PARA>();
+                        decryptPara.hProv = hKey;
+                        decryptPara.dwKeySpec = keySpec;
+                        decryptPara.pKeyAgree = pKeyAgreeRecipientInfo;
+                        decryptPara.dwRecipientIndex = pal.Index;
+                        decryptPara.dwRecipientEncryptedKeyIndex = pal.SubIndex;
+                        CMsgKeyAgreeOriginatorChoice originatorChoice = pKeyAgreeRecipientInfo->dwOriginatorChoice;
+                        switch (originatorChoice)
+                        {
+                            case CMsgKeyAgreeOriginatorChoice.CMSG_KEY_AGREE_ORIGINATOR_CERT:
+                                {
+                                    X509Certificate2Collection candidateCerts = new X509Certificate2Collection();
+                                    candidateCerts.AddRange(Helpers.GetStoreCertificates(StoreName.AddressBook, StoreLocation.CurrentUser, openExistingOnly: true));
+                                    candidateCerts.AddRange(Helpers.GetStoreCertificates(StoreName.AddressBook, StoreLocation.LocalMachine, openExistingOnly: true));
+                                    candidateCerts.AddRange(originatorCerts);
+                                    candidateCerts.AddRange(extraStore);
+                                    SubjectIdentifier originatorId = pKeyAgreeRecipientInfo->OriginatorCertId.ToSubjectIdentifier();
+                                    using (X509Certificate2 originatorCert = candidateCerts.TryFindMatchingCertificate(originatorId))
+                                    {
+                                        if (originatorCert == null)
+                                            return ErrorCode.CRYPT_E_NOT_FOUND.ToCryptographicException();
+                                        using (SafeCertContextHandle hCertContext = originatorCert.CreateCertContextHandle())
+                                        {
+                                            CERT_CONTEXT* pOriginatorCertContext = hCertContext.DangerousGetCertContext();
+                                            decryptPara.OriginatorPublicKey = pOriginatorCertContext->pCertInfo->SubjectPublicKeyInfo.PublicKey;
```

Comment:
(Or, alternatively, duplicate the CryptMsgControl call so we can do it while still inside this using block)

---
In reply to: [61309477](https://github.com/dotnet/corefx/pull/8083#discussion_r61309477) [](ancestors = 61309477)